Bundled tests

abadger opened the pull request:
##### SUMMARY
Add check that bundled code is up to date.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
Tests

##### ADDITIONAL INFORMATION
The release manager will turn this on in a PR before release and then open bugs for any software that is out of date.  Otherwise, this test will always be skipped.

abadger committed:
Have update-bundled check for updates to all bundled code

Add a test for whether we have bundled code inside of ansible that needs to be updated

mattclay commented on test/sanity/code-smell/update-bundled.py:
```diff
@@ -0,0 +1,104 @@
+
+        if LooseVersion(metadata['version']) < LooseVersion(latest_version):
+            needs_attention = True
+            print('UPDATE: {0} from {1} to {2} {3}'.format(
```
Messages should be in the format: `path: message`

Also, include a `update-bundled.json` file with:

```json
{
    "always": true,
    "output": "path-message"
}
```

That will tell ansible-test when to run the test and what format to expect messages in.

abadger replied:
Done.

mattclay commented on test/sanity/code-smell/update-bundled.py:
```diff
@@ -0,0 +1,104 @@
+                'https://pypi.org/pypi/{0}/json'.format(metadata['pypi_name'])))
+
+    if needs_attention:
+        sys.exit(1)
```
Exit codes are not necessary. Any messages will result in ansible-test considering the test failed.

abadger replied:
In my testing, the exit code was necessary.  Maybe it is because I don't have an update-bundled.json.  I'll add that file and test again.

abadger replied:
Yep, once the update-bundled.json metadata was added, the exit code was not needed.

Removed the sys.exit()

mattclay commented on test/sanity/code-smell/update-bundled.py:
```diff
@@ -0,0 +1,104 @@
+            metadata = get_bundled_metadata(filename)
+        except Exception as e:
+            needs_attention = True
+            print('WARNING: {0}'.format(e))
```
Warnings are not supported. This will result in an error. Print the error in the format: `path: message`

abadger replied:
Yeah, I do want it to result in an error for the test harness.  I'm not fussy about whether the message should read warning or error.

abadger replied:
Changed the message to read ERROR instead of WARNING.

mattclay commented on test/sanity/code-smell/update-bundled.py:
```diff
@@ -0,0 +1,104 @@
+                data = line[line.index('{'):].strip()
+                break
+        else:
+            raise Exception('{0} contained no metadata.  Could not check for updates'.format(filename))
```
Instead of raising and exception, print the error in the format: `path: message`

abadger replied:
This exception is caught further down.

abadger replied:
The catch is here: https://github.com/ansible/ansible/pull/51759/files#diff-36a6eef1b3c45cf98f385ac3a81b0cadR83

mattclay commented on test/sanity/code-smell/update-bundled.py:
```diff
@@ -0,0 +1,99 @@
+if __name__ == '__main__':
+    basedir = os.path.join(os.path.dirname(__file__), '../../..')
+
+    for filename in compat_libs(basedir):
```
Instead of hard-coding a list of files to check, why not check all `*.py` files for `_BUNDLED_METADATA` instead?

That will be easier to maintain and won't require updates to this test every time we add/remove a bundled file.

mattclay replied:
Well, all `*.py` files under any directories where we allow bundling anyways.

abadger replied:
I thought about that and I'm not sure.  The problem is a social problem and I can't figure out what technical implementations will be better or worse....

What we need is for all bundled code to be checked to make sure that it is up to date.  The problem is how do we detect that there is bundled code?

If we solely key off of the _BUNDLED_METADATA variable being present, then the test will stop detecting that a file needs to be updated if someone drops a new version in place and forgets that they have to add _BUNDLED_METADATA to the new file.

If we operate on files listed in this test case, then someone bundling new code needs to know that they have to add both _BUNDLED_METADATA to their file and list it here.

I think that hardcoding a list of files here is good to prevent the updated-version problem.  Maybe what we can do is to also scan for _BUNDLED_METADATA in files and if the two lists don't agree, make that an error as well.  That way failure to do the right thing in both places will lead to an error that can instruct users about what they forgot.

mattclay replied:
Checking both seems like the safest approach. It still won't catch cases of bundled code being added without _BUNDLED_METADATA -- but there probably isn't anything we can do in the sanity test about that.

abadger replied:
Okay, added that additional check

mattclay commented on test/sanity/code-smell/update-bundled.py:
```diff
@@ -23,27 +23,46 @@
+
+
+def get_files_with_bundled_metadata(basedir):
+    grep_cmd = subprocess.Popen(['grep', '-rwl', '--include=*.py', '_BUNDLED_METADATA', basedir],
```
Use os.walk to find files and search them in Python instead of using grep.

Although ansible-test currently doesn't support passing a list of all files (regardless of changes) to code-smell tests, that is something that is going to be fixed. Once fixed, tests will use that list instead of os.walk to discover files. This will allow ansible-test to handle skipping files so that individual tests don't need to re-implement the logic.

If we use grep now then it will need to be replaced later once the file list is available. Using os.walk instead will make applying the changes much easier.

abadger replied:
I'd rather use grep as using a list of files (generated by os.walk or by ansible-test) will require reinventing smarts that grep already has builtin and since it does it in C, it's almost certainly faster as well.

mattclay replied:
Using grep is almost certainly faster. However, if you're passing a list of files to grep you'll need to handle splitting up the file list into pieces small enough to avoid command line arg limits. Also, what grep "smarts" are you needing to use use?

abadger replied:
We don't have to pass a list of files to grep if we're operating on all files, right?  We just need to pass it the top of the tree.

In this case the smarts are word boundaries and include by file extension.  These are doable in python but I'm questioning what we gain by pushing the implementation of these things into our python code instead of letting grep handle those details?  It's not complexity and it's not speed but if there's another reason, then that's fine, just let me know what it is...

mattclay replied:
I've fixed the limitation in ansible-test that prevented code-smell tests from receiving file lists regardless of changed files. This test can now use the ansible-test provided file list and will not need to use os.walk or grep.

The `update-bundled.json` file should be changed as follows:

```json
{
    "ignore_changes": true,
    "extensions": [
        ".py"
    ],
    "output": "path-message"
}
```

The script only needs the following to get the list of files:

```python
paths = sys.argv[1:] or sys.stdin.read().splitlines()
```

As for the word boundary handling, a Python regex deals with that nicely:

```python
for path in paths:
    with open(path, 'r') as path_fd:
        contents = path_fd.read()

    if re.search(r'\b_BUNDLED_METADATA\b', contents):
        pass  # do something here
```

Use of the file list provided by ansible-test means that the script does not need to know anything about files or directories which should be skipped, as that logic already exists in ansible-test. For example, directories like .git, .tox, etc. are skipped. As the logic is changed/improved in ansible-test all of the code-smell tests benefit automatically. In many cases this can lead to improved performance since the test is checking far fewer files, even though the checks themselves may be a bit slower since they're in python.

Another benefit of avoiding grep is that we don't need to be worried about avoiding options that are not available in all versions of grep, since the logic is all in the python code.

If the test only needs to look at certain directories, ansible-test will filter those out as well. Here's an example configuration:

https://github.com/ansible/ansible/blob/06d83bae056bbc9de904a238e480acf1147fe80e/test/sanity/code-smell/no-main-display.json#L5-L8

abadger replied:
<nod>  Yep, a Python regex can do what grep does but it is still taking the things that grep has been optimized to do and attempting to write our own code that can only do it slower and with more complexity than shelling out.  But if portability is the overriding concern then I can make that change.

mattclay commented on test/sanity/code-smell/update-bundled.py:
```diff
@@ -0,0 +1,131 @@
+
+
+if __name__ == '__main__':
+    basedir = os.path.abspath(os.path.join(os.path.dirname(__file__), '../../..'))
```
This is not needed. The code-smell tests are always executed by ansible-test from the root of the source tree. Using basedir results in absolute file paths, but test results should be relative paths. Instead of making them absolute and converting them back, simply use relative paths instead.

abadger replied:
I thought about that but thought that glob.glob() would need a full path.  (Not sure why I thought that... maybe I was testing calling the script directly as well as through ansible-test).  Switching from glob.glob on the filesystem to operating on the paths passed in will remove the need for full paths as well.

abadger replied:
Done.

mattclay commented on test/sanity/code-smell/update-bundled.py:
```diff
@@ -0,0 +1,131 @@
+
+def get_bundled_libs(basedir):
+    bundled_libs = set()
+    for filename in glob.glob(os.path.join(basedir, 'lib/ansible/compat/*/__init__.py')):
```
Instead of scanning the filesystem again, the file list provided by ansible-test can be checked using fnmatch.

In the future ansible-test will include support for ignoring files not tracked by git. Using the file list will make sure this test correctly ignores those files.

abadger replied:
Makes sense.

abadger replied:
Done.

mattclay commented on test/sanity/code-smell/update-bundled.py:
```diff
@@ -0,0 +1,131 @@
+def get_bundled_libs(basedir):
+    bundled_libs = set()
+    for filename in glob.glob(os.path.join(basedir, 'lib/ansible/compat/*/__init__.py')):
+        if 'compat/tests' in filename:
```
Since there is no `compat/tests` in devel, I assume this is here so this test can be backported?

abadger replied:
I remember looking at compat/tests, seeing that it was now empty, but then I must have forgot to remove this later.  The script is very old, it's just that I'm getting around to automating it now (it lived in hacking before).

abadger replied:
Removed as there is no compat/tests anymore.

mattclay commented on test/sanity/code-smell/update-bundled.py:
```diff
@@ -0,0 +1,131 @@
+        filename = os.path.normpath(filename)
+        try:
+            metadata = get_bundled_metadata(filename)
+        except ValueError as e:
```
Include an exception handler for files that do not exist.

abadger replied:
Does this need to work on both Python2 and Python3?

abadger replied:
Used ```except (IOError, OSError)``` so that it will catch File not found on both Python2 and Python3.

abadger commented:
Damnit, github doesn't refresh the comments all the time.... I'll address the new comments before merging.

abadger committed:
Disable the update-bundled test during normal operation

This is enabled once per release cycle in a PR so that the release
manager can open bugs for the Ansible team to update the bundled code.

mattclay commented on test/sanity/code-smell/update-bundled.py:
```diff
@@ -0,0 +1,132 @@
+              ' test/sanity/code-smell/update-bundled.py'.format(filename))
+
+    for filename in bundled_libs:
+        filename = os.path.normpath(filename)
```
This is no longer necessary now that paths are simple relative paths.

abadger replied:
Done.

mattclay commented on test/sanity/code-smell/update-bundled.py:
```diff
@@ -0,0 +1,132 @@
+def get_files_with_bundled_metadata(paths):
+    with_metadata = set()
+    for path in paths:
+        if path == 'test/sanity/code-smell/update-bundled.py':
```
Are we allowing bundled files to be placed anywhere in the source tree?

If not, then the test will be more efficient if the directories allowed are added to the .json file as prefixes. Assuming `test/sanity/code-smell/` isn't one of those directories then this conditional could also be removed.

Example of using prefixes:

https://github.com/ansible/ansible/blob/06d83bae056bbc9de904a238e480acf1147fe80e/test/sanity/code-smell/no-main-display.json#L5-L8

abadger replied:
Yeah, they're allowed anywhere.  I half heartedly tried a couple times to specify a single directory (or two directories, one for controller-only and one for module_utils) but it never got enough traction from others so there's no rule. If someone would care to make some rules and move the code around then the test could be changed later.

mattclay reviewed (commented):
Looks pretty good now. Just a few minor comments you may want to address.

abadger commented:
Test failure on opensuse is unrelated.  The test is always going to be disabled inside of the devel branch anyway.  Merging so that it's ready when we branch.
